lib/mapper/custom_field_mapper.rb

```diff
@@ -2,25 +2,28 @@ module BigcommerceProductAgent
     module Mapper
         class CustomFieldMapper
 
-            def self.map(field_map, product, bc_product)
+            def self.map(field_map, raw_product, bc_product, current_custom_fields, namespace)
                 fields = {
                     upsert: [],
                     delete: [],
                 }
 
                 existing_fields = {}
 
-                if bc_product
-                    bc_product['custom_fields'].each do |cf|
+                if bc_product && current_custom_fields
+                    current_custom_fields.each do |cf|
                         cf['product_id'] = bc_product['id']
                         existing_fields[cf['name'].to_s] = cf
                     end
                 end
 
                 if field_map && field_map['additionalProperty']
                     field_map['additionalProperty'].each do |key, val|
-                        field = self.from_additional_property(product, existing_fields, key, val)
-                        fields[:upsert].push(field) unless field.nil?
+                        field = self.from_additional_property(raw_product, existing_fields, key, val)
```

Comment:
We should try to avoid the use of `key` and `value` in this context, because `value` isn't actually a value. Much like the method you're calling, `value` is really another key. It's a `from_{key|field|etc}` and a `to_{key|field|etc}`